File: btypes_primitives/src/main/js/de/hhu/stups/btypes/BBoolean.ts
import {BObject} from "./BObject";

export class BBoolean implements BObject{
    readonly value: boolean;

    constructor(value: boolean | string | BBoolean) {
        if (typeof (value) === 'boolean') {
            this.value = value;
        } else if (typeof value === 'string') {
            this.value = value.toLowerCase() === 'true';
        } else {
            this.value = value.value;
        }
    }
    
    and(other: BBoolean): BBoolean {
        return new BBoolean(this.value && other.value)
    }

    or(other: BBoolean): BBoolean {
        return new BBoolean(this.value || other.value)
    }

    not(): BBoolean {
        return new BBoolean(!this.value)
    }

    implies(other: BBoolean): BBoolean {
        return this.not().or(other)
    }

    equivalent(other: BBoolean): BBoolean {
        return this.implies(other).and(other.implies(this))
    }

    equal(other: BBoolean): BBoolean {
        return new BBoolean(this.value === other.value);
    }

    unequal(other: BBoolean): BBoolean {
        return new BBoolean(this.value != other.value);
    }

    booleanValue(): boolean {
        return this.value;
    }

    toString(): string {
        return this.value.toString();
    }

}

File: btypes_primitives/src/main/java/de/hhu/stups/btypes/BBoolean.java
package de.hhu.stups.btypes;

public final class BBoolean implements BObject {

	public static final BBoolean TRUE = new BBoolean(true);
	public static final BBoolean FALSE = new BBoolean(false);

	private final boolean value;

	public static BBoolean of(boolean value) {
		if (value) {
			return TRUE;
		} else {
			return FALSE;
		}
	}

	public static BBoolean of(String value) {
		return of(Boolean.parseBoolean(value));
	}

	public BBoolean(boolean value) {
		this.value = value;
	}

	public BBoolean(String s) {
		this(Boolean.parseBoolean(s));
	}

	public boolean booleanValue() {
		return this.value;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) {
			return true;
		} else if (!(o instanceof BBoolean)) {
			return false;
		} else {
			return this.value == ((BBoolean) o).value;
		}
	}

	@Override
	public int hashCode() {
		return Boolean.hashCode(this.value);
	}

	@Override
	public String toString() {
		return String.valueOf(this.value);
	}

	/* groovy operator overloading support */
	Object asType(Class<?> clazz) {
		if (clazz == BBoolean.class) {
			return this;
		}
		return clazz.cast(this.value);
	}

	public BBoolean or(BBoolean other) {
		return new BBoolean(this.booleanValue() || other.booleanValue());
	}

	public BBoolean or(boolean other) {
		return new BBoolean(this.booleanValue() || other);
	}

	public BBoolean xor(BBoolean other) {
		return new BBoolean(this.booleanValue() ^ other.booleanValue());
	}

	public BBoolean xor(boolean other) {
		return new BBoolean(this.booleanValue() ^ other);
	}

	public BBoolean and(BBoolean other) {
		return new BBoolean(this.booleanValue() && other.booleanValue());
	}

	public BBoolean and(boolean other) {
		return new BBoolean(this.booleanValue() && other);
	}

	public BBoolean not() {
		return new BBoolean(!this.booleanValue());
	}

	public BBoolean implies(BBoolean other) {
		return this.not().or(other);
	}

	public BBoolean implies(boolean other) {
		return new BBoolean(this.not().booleanValue() || other);
	}

	public BBoolean equivalent(BBoolean other) {
		return new BBoolean(this.booleanValue() == other.booleanValue());
	}

	public BBoolean equivalent(boolean other) {
		return new BBoolean(this.booleanValue() == other);
	}

	public BBoolean equal(BBoolean other) {
		return new BBoolean(this.booleanValue() == other.booleanValue());
	}

	public BBoolean unequal(BBoolean other) {
		return new BBoolean(this.booleanValue() != other.booleanValue());
	}

	public BBoolean isBoolean() {
		return TRUE;
	}

	public BBoolean isNotBoolean() {
		return FALSE;
	}
}
